[FEATURE] [MER-3010] Activity reports display component

rgachuhi opened the pull request:
This PR adds a new feature to Torus that allows any activity to generate a report, which can be displayed on any other page or page via the new "Report" component. 

Example: The legacy student cognition toolbox uses a component in which students take a survey on one page and then will see their results on the next. This is called a survey report in legacy.

https://github.com/Simon-Initiative/oli-torus/assets/11138398/b5b2ffbb-0e5c-46e5-816d-921dbff90355


https://github.com/Simon-Initiative/oli-torus/assets/11138398/acd3a744-31b8-44c2-bf03-d90a2527fb80

eliknebel commented on lib/oli/rendering/report.ex:
```diff
@@ -0,0 +1,34 @@
+
+  @callback report(%Context{}, %{}) :: [any()]
+  @callback error(%Context{}, %{}, {Atom.t(), String.t(), String.t()}) :: [any()]
+
```
`Atom.t()` shows as an unknown type. Use `atom()` instead

```suggestion
  @callback error(%Context{}, %{}, {atom(), String.t(), String.t()}) :: [any()]
```

rgachuhi replied:
done

eliknebel commented on assets/src/components/resource/editors/createEditor.tsx:
```diff
@@ -58,6 +59,9 @@ export const createEditor = (editorProps: EditorProps): JSX.Element => {
     case 'survey':
       return <SurveyEditor {...editorProps} contentItem={contentItem} />;
+    case 'report':
+      console.log('creating report editor');
```
Remove console log

rgachuhi replied:
Removed

eliknebel reviewed (commented):
Implementation looks good, tested and works as expected. Just a few comments about databse calls in the rendering codepath and console log statements

rgachuhi commented on assets/src/components/misc/VegaLiteRenderer.tsx:
```diff
@@ -0,0 +1,16 @@
+}
+
+export const VegaLiteRenderer = (props: VegaLiteSpec) => {
+  console.log(' sdhiere');
```
Done

rgachuhi commented on lib/oli/rendering/report/html.ex:
```diff
@@ -0,0 +1,113 @@
+    pub_ids = DeliveryResolver.section_publication_ids(section_slug) |> Oli.Repo.all()
+
+    Enum.reduce(pub_ids, %{}, fn a, c ->
+      case Map.get(Oli.Publishing.determine_parent_pages([activity_id], a), activity_id) do
```
I could not find a sensible way to precompute the function, but I have refactored it to be a more efficient query

darrensiegel commented on lib/oli/delivery/attempts/core.ex:
```diff
@@ -589,6 +589,26 @@ defmodule Oli.Delivery.Attempts.Core do
   end
 
+  def get_latest_activity_attempt(section_id, user_id, resource_id) do
+    Repo.one(
```
This does not take into account that a resource (page) can have multiple attempts.  This should be reworked to ensure that you are fetching the latest attempt for this activity WITHIN the latest attempt for the page.

rgachuhi replied:
Not sure I understand this. From what I tell, the query fetches the last activity attempt based on the greater activity attempt id, which means that for a particular activity, user and section combo, it will get the greatest activity attempt regardless of what page attempt that belongs to.

darrensiegel commented on lib/oli/activities/reports/providers/OliLikert.ex:
```diff
@@ -0,0 +1,166 @@
+      Enum.reduce(activity_attempt.part_attempts, %{}, fn a, c -> Map.put(c, a.part_id, a) end)
+
+    data =
+      Map.get(activity_attempt.revision.content, "items")
```
The `revision` field should not be directly used here, as doing so would not take into account dynamic questions.  Instead, one must select the model, taking into account the `transformed_model` attribute of the activity attempt.  If it is not `nil`, use that.  If it is `nil`, using the content of the `revision`.

rgachuhi replied:
Issue addessed

darrensiegel commented on lib/oli/publishing.ex:
```diff
@@ -1489,7 +1489,8 @@ defmodule Oli.Publishing do
   page that encloses it
   """
-  def determine_parent_pages(activity_resource_ids, publication_id) do
+  def determine_parent_pages(activity_resource_ids, publication_ids)
```
List support is added here, but it doesn't seem to be used anywhere.  If it isn't being used, we should revert these changes.

rgachuhi replied:
That is used in the `defp determine_parent_link(section_slug, activity_id) do` at `defmodule Oli.Rendering.Report.Html do`

darrensiegel reviewed (changes requested):
This looks really good.  There are a few things that have to be addressed.  After that is done I'll do a second pass review where I'll test this out.

darrensiegel commented:
Also, I see in the screenshot that the VegaLite actions menu is present. We want to disable that: https://stackoverflow.com/questions/63715640/how-to-hide-the-three-dot-dropdown-menu-on-a-vega-chart

rgachuhi commented:
> <img alt="Screenshot 2024-06-11 at 7 28 53 PM" width="975" src="https://private-user-images.githubusercontent.com/6248894/338990138-2eb23490-a2b7-4c04-9698-cd8b3ce8d219.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyODYxMTUsIm5iZiI6MTcxODI4NTgxNSwicGF0aCI6Ii82MjQ4ODk0LzMzODk5MDEzOC0yZWIyMzQ5MC1hMmI3LTRjMDQtOTY5OC1jZDhiM2NlOGQyMTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTNUMTMzNjU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2U1YTExMTYxNDQ1MjI5ZmNmNjI1M2ZkMzFkOGMwZTdiOWE4MmYwOGUzZTRiN2E5N2NjNGQ3Mzc4NGMyNjk1YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.CrM2lB5OgsZtUocHUiN2d6DpzgLW_Svj6fBNzA6GZho">
> I noticed that Vegalite has a menu that allows you to open in an editor. Is there any way to disable this menu?
> 
> Also, is this number scale intentional? The numbers seem unnecessarily large

Not sure why the ranges are that high, but it all depends on the values the likert activity sets

rgachuhi commented:
> Also, I see in the screenshot that the VegaLite actions menu is present. We want to disable that: https://stackoverflow.com/questions/63715640/how-to-hide-the-three-dot-dropdown-menu-on-a-vega-chart

I was able to disable those.

rgachuhi commented on assets/src/components/misc/VegaLiteRenderer.tsx:
```diff
@@ -0,0 +1,16 @@
+}
+
+export const VegaLiteRenderer = (props: VegaLiteSpec) => {
+  console.log(' sdhiere');
```
Removed

darrensiegel commented on lib/oli/activities/reports/providers/OliLikert.ex:
```diff
@@ -0,0 +1,172 @@
+      ) do
+    with %{choices: choices, data: data} <-
+           process_attempt_data(enrollment.section_id, user.id, activity_id) do
+      data_url = "/api/v1/activity/report/#{enrollment.section_id}/#{activity_id}"
```
We shouldn't hardcode URLs in code.  Instead, use the `Routes` module to determine the url.  But doing so here would lead to a dependency from `Oli` to `OliWeb`, which we don't want.   So you will need to pass to this function a function that can build the URL.

darrensiegel replied:
The purpose of splitting the data apart from the visualization and instead embedding a URL to the controller was so that we do not have to do any DB calls during the rendering phase.  But `process_attempt_data` ends up making three DB queries.  One for the "get_latest_attempt", then two for the external `preloads`.  

We need to remove all of these DB calls here.

rgachuhi replied:
No hardcoded URL in the backend. Moved logic into a new likert frontend component

I re-implemented it with a new solution that avoids most db calls except for a minor one on the render function.

darrensiegel commented on lib/oli/activities/reports/providers/OliLikert.ex:
```diff
@@ -0,0 +1,172 @@
+        c ++ [value]
+      end)
+
+    %{choices: Map.get(activity_attempt.revision.content, "choices"), data: data}
```
We shouldn't use the revision directly, as this will not honor dynamic question content.

rgachuhi replied:
Changed to honor the transformed model and fallback to content if model not transformed.

darrensiegel commented on lib/oli/delivery/attempts/core.ex:
```diff
@@ -589,6 +589,26 @@ defmodule Oli.Delivery.Attempts.Core do
+        join: ra in ResourceAttempt,
+        on: ra.id == aa.resource_attempt_id,
+        join: rac in ResourceAccess,
+        on: rac.id == ra.resource_access_id,
```
Maybe this actually should be querying to find the latest "evaluated" activity?

rgachuhi replied:
Latest "evaluated" activity check added to the query

darrensiegel reviewed (changes requested):
Two things to change and a question around the "get_latest_activity_attempt". 

Also, perhaps I am just mis-remembering, but I thought that the legacy OLI activity report showed the student not "their" response, but instead the aggregation of their classmates responses for the activity in question?  As in "Here is what your peers answered to this question".

darrensiegel commented:
I tested this out locally with a new course. I first created a practice page and added a Likert activity.  It contained three choices "Agree", "Neither Agree or Disagree" and "Disagree".  I added two Items "Item one" and "Item two".  I then added an activity report to a second page, published and accessed the course as a student. 

Visiting the page with the report first correctly showed me the message that I hadn't answered the question on the other page. 

After answering the question "Disagree" for Item 1 and "Agree" for item 2, then clicking "Submit" for the Likert, I went back to the page with the report and see this:
<img width="989" alt="Screenshot 2024-06-14 at 7 56 29 AM" src="https://github.com/Simon-Initiative/oli-torus/assets/7431756/74bd286d-3580-4ef6-aec0-9c78320b9de8">

The visualization doesn't seem to be correct in a few ways.  First, it isn't showing all choices in the Y axis, only the two that I picked.  Second, the Y axis scale doesn't make sense.  And finally, I'm not sure what good using a horizontal bar chart here is.  It immediately gives me the sense that P1 is "more" or "better" than P2.    Much better would be something along the lines of https://vega.github.io/vega-lite/examples/layer_likert.html

But as I also pointed out in the other comment, I thought that the Legacy impl of this report showed the "Aggregate" results of this activity across all students in the course?  

I then reset the activity on the page and went back to the page with the report and now see this:
<img width="989" alt="Screenshot 2024-06-14 at 7 57 46 AM" src="https://github.com/Simon-Initiative/oli-torus/assets/7431756/88f12125-8e44-49e8-804c-28e8377e18c9">

So I think we probably should be only querying for the latest "evaluated" activity

rgachuhi commented:
> I tested this out locally with a new course. I first created a practice page and added a Likert activity. It contained three choices "Agree", "Neither Agree or Disagree" and "Disagree". I added two Items "Item one" and "Item two". I then added an activity report to a second page, published and accessed the course as a student.
> 
> Visiting the page with the report first correctly showed me the message that I hadn't answered the question on the other page.
> 
> After answering the question "Disagree" for Item 1 and "Agree" for item 2, then clicking "Submit" for the Likert, I went back to the page with the report and see this: <img alt="Screenshot 2024-06-14 at 7 56 29 AM" width="989" src="https://private-user-images.githubusercontent.com/7431756/339759419-74bd286d-3580-4ef6-aec0-9c78320b9de8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkyNDM0MzcsIm5iZiI6MTcxOTI0MzEzNywicGF0aCI6Ii83NDMxNzU2LzMzOTc1OTQxOS03NGJkMjg2ZC0zNTgwLTRlZjYtYWVjMC05Yzc4MzIwYjlkZTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjRUMTUzMjE3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODllYzEzNzhhZThmZGI1NjJlZDhlYzAzNjBmNDllNDFlYTExYTkzOTlmMTE4NjViNjYwYmRkMjgyNjUzY2RhMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.iwqsIEnGJO_o3kZu6hAA9EVK7JN5vnl4y-qoE5QD3tE">
> 
> The visualization doesn't seem to be correct in a few ways. First, it isn't showing all choices in the Y axis, only the two that I picked. Second, the Y axis scale doesn't make sense. And finally, I'm not sure what good using a horizontal bar chart here is. It immediately gives me the sense that P1 is "more" or "better" than P2. Much better would be something along the lines of https://vega.github.io/vega-lite/examples/layer_likert.html
> 
> But as I also pointed out in the other comment, I thought that the Legacy impl of this report showed the "Aggregate" results of this activity across all students in the course?
> 
> I then reset the activity on the page and went back to the page with the report and now see this: <img alt="Screenshot 2024-06-14 at 7 57 46 AM" width="989" src="https://private-user-images.githubusercontent.com/7431756/339760490-88f12125-8e44-49e8-804c-28e8377e18c9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkyNDM0MzcsIm5iZiI6MTcxOTI0MzEzNywicGF0aCI6Ii83NDMxNzU2LzMzOTc2MDQ5MC04OGYxMjEyNS04ZTQ0LTQ5ZTgtODA0Yy0yOGU4Mzc3ZTE4YzkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjRUMTUzMjE3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjdiYTk3MzkyYzY0NTljNDE0ZmZhNzUwMDIwOTRiMTEzNDYwYmY1Y2I5ZDRhODFkYzg2YzRmMjdlOGE3YzY0MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.xITP04xV83DEWxrXuXpFuPG8-UyS9DqZFN98b6tXc2E">
> 
> So I think we probably should be only querying for the latest "evaluated" activity

I have addressed the three issues commented on above. One; improved the visualization to something that makes more sense, two; ensured that the likert scalling works correctly, three; ensured that the last "evaluated" attempt is respected

The legacy impl is indeed reporting for individual users not the "Aggregate" :-)

rgachuhi commented:
> <img alt="Screenshot 2024-06-11 at 7 28 53 PM" width="975" src="https://private-user-images.githubusercontent.com/6248894/338990138-2eb23490-a2b7-4c04-9698-cd8b3ce8d219.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkyNDU2MzYsIm5iZiI6MTcxOTI0NTMzNiwicGF0aCI6Ii82MjQ4ODk0LzMzODk5MDEzOC0yZWIyMzQ5MC1hMmI3LTRjMDQtOTY5OC1jZDhiM2NlOGQyMTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjRUMTYwODU2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTVjNTMyYzFiZGFiOGE1MDg5NzQwMGNjYzRjYzUwZWQ2NTA1ZjBmMzRhNWM5NWU2NGViZmM0OGVlYzAxZjQ1YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.DF8MvhQJbnae77P3Mo-LHoe4xUTK7tKLsThhT8-nq5Q">
> I noticed that Vegalite has a menu that allows you to open in an editor. Is there any way to disable this menu?
> 
> Also, is this number scale intentional? The numbers seem unnecessarily large

darrensiegel commented on assets/src/components/activities/likert/reports/LikerReportRenderer.tsx:
```diff
@@ -0,0 +1,80 @@
+  parent: Parent;
+};
+
+export const LikerReportRenderer = (props: LikerReportRendererProps) => {
```
I just noticed this is misspelled `Liker` instead of `Likert`

rgachuhi replied:
I shifted all DB work from the `Oli.Rendering.Report.Html` report function. That way, rendering of any page containing the report is fast. The actual contents of the activity are rendered when the client-side LikertReportRenderer calls back to the `activity_report_data_controller` to fetch data in addition to other content that requires DB calls. I figured that I may as well bundle the data for the VegaLite with the spec for this activity report. Other activities may choose to use the `activity_report_data_controller` otherwise.

darrensiegel reviewed (changes requested):
One minor thing with a incorrect spelling. 

Also, it appears as all of the DB work is now done during rendering, but it looks like the activity_report_data_controller and the LikerReportRenderer component are still in place.  Can we now delete those?
